relay-cardinality/src/limiter.rs

```diff
@@ -0,0 +1,394 @@
+//! Relay Cardinality Limiter
+
+use std::collections::BTreeSet;
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+use relay_metrics::{Bucket, MetricNamespace, MetricResourceIdentifier};
+
+use hash32::{FnvHasher, Hasher as _};
+use relay_statsd::metric;
+
+use crate::statsd::CardinalityLimiterTimers;
+use crate::{OrganizationId, Result};
+
+/// Limiter responsible to enforce limits.
+pub trait Limiter {
+    /// Verifies cardinality limits.
+    ///
+    /// Returns an iterator containing only accepted entries.
+    fn check_cardinality_limits<I, T>(
+        &self,
+        organization: OrganizationId,
+        entries: I,
+        limit: usize,
+    ) -> Result<Box<dyn Iterator<Item = Rejection>>>
+    where
+        I: IntoIterator<Item = Entry<T>>,
+        T: CardinalityScope;
+}
+
+/// The scope of a [`CardinalityItem`].
+pub trait CardinalityScope: Debug + Display + Hash + PartialEq + Eq + 'static {}
+
+impl<T> CardinalityScope for T where T: Debug + Display + Hash + PartialEq + Eq + 'static {}
+
+/// Unit of operation for the cardinality limiter.
+pub trait CardinalityItem {
+    /// The cardinality limit scope.
+    type Scope: CardinalityScope;
+
+    /// Transforms this item into a consistent hash.
+    fn to_hash(&self) -> u32;
+
+    /// Extracts a scope to check cardinality limits for from this item.
+    fn to_scope(&self) -> Option<Self::Scope>;
+}
+
+// Implementation for a bucket, currently here, may also be implemented in `relay-metrics`, if it
+// makes sense to move it.
+impl CardinalityItem for Bucket {
```

Comment:
Yeah, having this implemented in metrics in my opinion would be cleaner.  Since relay-metrics would be responsible  to provide the implementation and this crate just provides the facilities for the entire thing to work.

Reply:
I'll move it!